### PR TITLE
Add vllm chat completion endpoints and fix inspect_history

### PR DESCRIPTION
**Pull request by noobHappylife**
Related to issue #778 #784 

- added vllm chat completion endpoints (support system_prompt)
- fixed an issue with .copy() with vllm client module. (need to include model, url, port into kwargs)
- fix inspect_histrory not respecting n/skip as mentioned in #784

### CODE REVIEW

File: dsp/modules/hf_client.py [hunk 1]
```diff
@@ -128,47 +128,89 @@ def __init__(self, model, port, url="http://localhost", **kwargs):
         
         self.headers = {"Content-Type": "application/json"}
         self.kwargs |= kwargs
+        # kwargs needs to have model, port and url for the lm.copy() to work properly
+        self.kwargs.update({
+            'model': model,
```
Comment (arnavsinghvi11) [inline dsp/modules/hf_client.py:133]: model is in fact included in the kwargs already, inherited through [HFModel-> LM](https://github.com/stanfordnlp/dspy/blob/0b7289fca05988b2863cc46a3564e8edbc9b7021/dsp/modules/lm.py#L7) so we can remove this.

is there a use case for including the port and URL in the LM kwargs? seems like they are only used for setting the URL(s) and not in the model payload? feel free to remove if not.

File: dsp/modules/hf_client.py [hunk 2]
```diff
@@ -128,47 +128,89 @@ def __init__(self, model, port, url="http://localhost", **kwargs):
...
-
-        payload = {
-            "model": self.kwargs["model"],
-            "prompt": prompt,
-            **kwargs,
-        }
-
         
+        # get model_type
+        model_type = kwargs.get("model_type",None)
```
Comment (arnavsinghvi11) [inline dsp/modules/hf_client.py:143]: let's actually move this to the initialization layer so the user can specify the model_type when configuring the vLLM client, not during generation time.

File: dsp/modules/lm.py [hunk 3]
```diff
@@ -63,13 +63,11 @@ def inspect_history(self, n: int = 1, skip: int = 0):
             if len(printed) >= n:
                 break
 
-        for idx, (prompt, choices) in enumerate(reversed(printed)):
-            printing_value = ""
-
+        for idx, (prompt, choices) in enumerate(printed):
```
Comment (arnavsinghvi11) [inline dsp/modules/lm.py:66]: is there a reason for this change? seems to have the same effect originally?

**Review by arnavsinghvi11 — CHANGES_REQUESTED** [review]

File: dsp/modules/lm.py (hunk 3 above)
Comment (arnavsinghvi11) [reply dsp/modules/lm.py:66]: currently failing these [tests](https://github.com/stanfordnlp/dspy/actions/runs/8671183621/job/23806529670?pr=811)

**Comment by arnavsinghvi11** [comment]
Thanks @noobHappylife  for the additions to vLLM. left a couple comments and should be good to merge following those comments and running `ruff check . --fix-only` .

File: dsp/modules/hf_client.py (hunk 1 above)
Comment (noobHappylife) [reply dsp/modules/hf_client.py:133]: Noted on the model part. For url and port, basically is when calling lm.copy() (ex: while compiling a program with bootstrap), it is missing the url and port kwargs. As shown as below. Thus, requiring the port and url in the kwargs for the copy() to work properly. 

```
in BootstrapFewShot._bootstrap_one_example(self, example, round_idx)
    148 with dsp.settings.context(trace=[], **self.teacher_settings):
    149     lm = dsp.settings.lm
--> 150     lm = lm.copy(temperature=0.7 + 0.001 * round_idx) if round_idx > 0 else lm
    151     new_settings = dict(lm=lm) if round_idx > 0 else {}
    153     with dsp.settings.context(**new_settings):

File [~/micromamba/envs/yh_hf/lib/python3.10/site-packages/dsp/modules/lm.py:108](http://192.168.77.2:8892/lab/tree/~/micromamba/envs/yh_hf/lib/python3.10/site-packages/dsp/modules/lm.py#line=107), in LM.copy(self, **kwargs)
    105 kwargs = {**self.kwargs, **kwargs}
    106 model = kwargs.pop("model")
--> 108 return self.__class__(model=model, **kwargs)

TypeError: HFClientVLLM.__init__() missing 1 required positional argument: 'port'
```

File: dsp/modules/lm.py (hunk 3 above)
Comment (noobHappylife) [reply dsp/modules/lm.py:66]: I've included some of my testing with the inspect_history in the issue #784. Can you take a look, or I'm misunderstanding how it should works?

Comment (arnavsinghvi11) [reply dsp/modules/lm.py:67]: Ah thanks for highlighting. Just realized that this is a bug where the `printing_value` is getting reset everytime in the loop. That should be all the error is and the skip condition is fine as is. Adding a commit to this file - lmk if it aligns with fixing your issue and will merge.

Comment (noobHappylife) [reply dsp/modules/lm.py:67]: Thank you. Let me know what else is needed to change. I've updated to the vllm initialization part, and fix the error, so the test should pass now.

Comment (arnavsinghvi11) [reply dsp/modules/lm.py:67]: yeah can you revert the changes made to `inspect_history` here and just move the [`printing_value`](https://github.com/stanfordnlp/dspy/pull/811/files#diff-6bdc17664251e0662d9b19a69a5ebd5acd2252e4ecde2bdb6d89c460b1d022a6L67) outside the for loop?

Comment (noobHappylife) [reply dsp/modules/lm.py]: ok, updated and tested it works. Thank you. Now inspect_history is able to view n-number of histories with skip.

Comment (ammirsm) [reply dsp/modules/lm.py]: good catch folks! 

thanks for it!

**Comment by arnavsinghvi11** [comment]
Thanks @noobHappylife !

**Pull request merged**